[Spring-Ecommerce-Order Step 1] Submit miooo0o's Mission

miooo0o opened the pull request:
Hello Dani, thank you for understanding the late submission.
I’m sorry that it ended up being delayed by another day as well. 🙇 

## migrate from ecommerce and complete JPA refactor

This PR initializes the `spring-ecommerce-order` project based on the existing `spring-ecommerce` project from my pair programming partner and completes the implementation of core order-related features along with a comprehensive refactor.

---

## Step 1.1 – Product API

### Entity and Relationship Setup
- Created `Product`, `Member`, `Cart`, `CartItem`, and `Option` entities and defined their relationships
- Improved handling of `CartItem.createdAt` and `updatedAt`
### Repository Structure Changes
- Migrated all repositories (including `Product` and `Member`) to JPA
- Removed JDBC dependencies

**Key decision:** We first refactored the JDBC-based repositories to match JPA return types, then fully migrated them to JPA.  
**Key decision:** For `Cart` and `CartItem`, DTOs were being used as entities, and modifying tables to fit JDBC looked too complex. We decided to delete the old implementation and rewrite it from scratch instead, which also impacted the tests.

## Step 1.2 – Pagination

### Feature Implementation
- **Cart Controller & Service:** Added item add/remove functionality and pagination
- **Product Controller & Service:** Added pagination and option management (Entity + DTO)
- **Statistics:** Fully refactored to JPA and improved sorting conditions

## Step 1.3 – Product Option
**Key decision:** We chose not to handle `Option` separately but instead keep it as a property of `Product`. and also use `CascadeType.ALL`.

```kotlin
@Entity
@Table(name = "products")
class Product(
    ...
) {
    @OneToMany(mappedBy = "product", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    var options: MutableList<Option> = mutableListOf()
}
```

- Reason: To minimize changes to the existing code base due to time constraints

---

## Test Improvements

- Added controller tests for `Product`, `Member`, and `Cart` using MockMvc
- Introduced a `Fixture` object for reusable test instances
- Refactored statistics tests to use JPA
- Added unit tests for `Cart` service and controller

**Key decision:** We encountered an issue where updates to a `Product`'s `Option` list were not properly persisted to the database, especially in tests that used lateinit. To resolve this, we used `@Transactional` and explicitly accessed the options collection to force initialization.
```kotlin
@Transactional
@Service
class ProductService(private val productRepository: ProductRepository) {
    ...
    fun read(): List<Product> {
        val products = productRepository.findAll()
        products.forEach { it.options.size } // call all options here
        return products
    }
}
```
- Reason: To minimize changes to the existing code base due to time constraints


---

Thank you so much for reviewing this.
Let me know if anything is unclear or needs improvement and again, thank you for your understanding!

miooo0o committed:
init files from ecommerce

miooo0o committed:
init( build.gradle.kts, application.properties ): add dependencies

miooo0o committed:
docs( README ): update readme

miooo0o committed:
refactor( Product.kt ): transfer to entity and add annotation

miooo0o committed:
refactor( Product.kt, application.properties ): match schema table columns with entity and add a property so that Hibernate does not recreate tables

BREAKING CHANGE: tests run

miooo0o committed:
refactor( Member.kt ): transfer to entity and add annotation

- rename toMemberDto to toRegisteredMember
- now Member.role defined as Role.**.name

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor( CartItem.kt, CartItemResponse.kt ): change return type in repository add/delete methods

- to apply association Product filed to CartItem entity, reduce CartItemResponse usage
- applied change to service/controller

miooo0o committed:
test: mute tests except Product, Member controller

- Decision: keep Product, Member but redesign Cart, CartItem

miooo0o committed:
refactor( CartItem.kt, CartItemResponse.kt ):

- Decision: if re-design, we decide to use `id` only depend on level
- these classes are still in progress

miooo0o committed:
refactor( CartItem.kt, Cart.kt ): change table design and entity, implement related test `CartEntityTest` with `CartRepositoryJPA`

- "spring.jpa.hibernate.ddl-auto=create-drop" to check sql table
- class CartEntityTest
- interface CartRepositoryJPA

miooo0o committed:
chore: cleanup coe

petrabenczedh committed:
refactor( model ): change entity relationships and table design

petrabenczedh committed:
docs( README ): update feature list

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor/feat( ProductRepository.kt ): use interfaces to refactor JDBC repo to JPA repo

- replace JDBC with JPA inside the ProductControllerTest.kt
- refactor functions inside the service and repository for Product

miooo0o committed:
chore: mute test

miooo0o committed:
refactor/feat( MemberRepository.kt, TokenLoginControllerTest.kt ): use interfaces to refactor JDBC repo to JPA repo

- replace JDBC with JPA inside the TokenLoginControllerTest.kt
- refactor functions inside the service and repository for Member

miooo0o committed:
chore: cleanup code

miooo0o committed:
chore: cleanup code

miooo0o committed:
Merge remote-tracking branch 'origin/step1' into step1

miooo0o committed:
refactor( CartItem ): change type of quantity to Int from Long

miooo0o committed:
chore: cleanup code

- remove unused JDBC-related files and lines
- adjust schema.sql conditions (not used with JPA)
- cleanup application.properties

miooo0o committed:
test: add a Fixture object to hold pre-made test instances

miooo0o committed:
test(service): add a new test to Cart Service

miooo0o committed:
refactor(cart): add controller with minimal service and repository for Cart/CartItem

- still need to implement

miooo0o committed:
refactor( Cart.kt ): add business logic to addItem and removeItem methods

miooo0o committed:
feat( DatabaseConfig.kt, DatabaseFixture.kt ): predefine table values for the application and for testing

- what previously has been data.sql's job is now done here
- contains constants

miooo0o committed:
style: run ktlint formatter

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(cart): extend entity methods and adjust DTO parameters for usage

miooo0o committed:
feat(cart): add add/delete endpoints to controller and service layer

miooo0o committed:
test: implement cart controller and service basic tests

miooo0o committed:
chore: remove unused DatabaseConfig.kt

miooo0o committed:
test(cart): refactor fixture setup into reusable function

- Extract fixture creation into a helper function

miooo0o committed:
test: merge exception tests to a controller test file and separate package

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(stat): apply JPA to stat repository

petrabenczedh committed:
refactor( TokenLoginControllerTest.kt ): remove DIRTIESCONTEXT and make test work with Fixtures

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): add random port instead of the defined port, so that tests can run independently

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): move to other folder

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): start refactoring StatisticsE2ETest.kt

miooo0o committed:
feat( ProductController, ProductService): implement pagination for the product list

miooo0o committed:
test(ProductControllerMock): add new test to `getPages` method with MockMvc

miooo0o committed:
test(CartController): add new test to `getPages` method with MockMvc

miooo0o committed:
feat( CartController, CartService): implement pagination for the Cart list as CartItemResponse

miooo0o committed:
refactor( ProductController ): rename getPages to asPages

miooo0o committed:
feat( Option.kt ); add Option Entity and Transactional annotations

- products.forEach { println(it.options) }
is used to load options lazily

miooo0o committed:
feat( OptionRequest/Response ): implement new DTOs to add options filed in Product and apply changed to related layers

miooo0o committed:
test(product): add new test to check OptionResponse to ProductController

- create new method to create Product with Option

miooo0o committed:
chore: comment in `spring.jpa.open-in-view=false` and remove @Disabled annotation from CartEntityTest

miooo0o committed:
chore: cleanup code

- remove unused methods/properties general

miooo0o committed:
test(product-controller): refactor to use OptionRepository for option tracking

miooo0o committed:
chore: remove TODO comments for a final version

- This commit removes leftover TODO comments to prepare for the final release.
- Remaining parts that may need fixes can be identified from this commit.

miooo0o committed:
test(stat): refactor test to apply JPA

miooo0o committed:
test(product-controller): create() test doesn't check created product id

miooo0o committed:
fix(stat-service): change sort condition check `updatedAt` in CartItem

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(cart, cart-item): change CartItem update `createdAt` and `updatedAt` menially

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
refactor(product-form): rename class to ProductForm from ProductFormDto

miooo0o committed:
refactor(product, option): replace mutable vars with immutable vals where reassignment is not needed

miooo0o committed:
refactor(cart-item, option): add quantity update logic and centralize entity constants

- Added logic to update `quantity` (currently unused, prepared for future order feature)
- Moved all entity-related constants into companion objects for better organization

miooo0o committed:
refactor(cart): split large functions into smaller ones

miooo0o committed:
refactor(option-response): rename id to optionId

da-nyee commented on src/main/kotlin/ecommerce/model/Product.kt:
```diff
@@ -0,0 +1,45 @@
+
+//    init {
+//        require(options.isNotEmpty()) { "Product name must not be blank" }
+//    }
```
What about removing these comments if they're unnecessary?

miooo0o replied:
yeah, I completely missed! I would double check this time.

da-nyee commented on src/main/kotlin/ecommerce/model/Option.kt:
```diff
@@ -0,0 +1,45 @@
+    @Column(nullable = false)
+    var quantity: Int = 1,
+    @ManyToOne(fetch = FetchType.EAGER)
+    var product: Product,
```
Is there any particular reason you set this as `EAGER`?

miooo0o replied:
Maybe an error occurred when `product` was set to `LAZY`? I had been using `FetchType` without really understanding it, so I had no clear reason for the choice mostly just wanted to avoid some errors. 

After studying it today, I switched it to `LAZY`.

__

According to what I learned today, `EAGER` fetching carries the risk of loading large amounts of unnecessary data, which can hurt performance. In practice, `LAZY` fetching combined with `fetch joins` is more commonly used, and even `EAGER` fetching cannot fully prevent the `N+1 problem` in some cases.

Am I understanding this correctly?

da-nyee replied:
Yes! You've got it 👍

da-nyee commented on src/main/kotlin/ecommerce/model/Option.kt:
```diff
@@ -0,0 +1,45 @@
+    val id: Long = 0L,
+) {
+    init {
+        require(quantity in 1 until 100_000_000) { "quantity must be in range 1..100_000_000" }
```
Underscore 👍

da-nyee commented on src/main/kotlin/ecommerce/model/Option.kt:
```diff
@@ -0,0 +1,45 @@
+        require(quantity in 1 until 100_000_000) { "quantity must be in range 1..100_000_000" }
+        require(name.length <= 50) { "name length must be <= 50 characters" }
+        require(name.all { it.isLetterOrDigit() || it in allowedSpecialChars })
+        if (product.options.isNotEmpty()) {
+            require(product.options.all { it.name != this.name }) { "duplicate name ${product.name} found" }
+        }
```
I'd recommend delegating this validation to the service layer,
since it seems like business logic, and fetching other options for a product within an option feels a bit awkward 🤔

miooo0o replied:
While trying to fix this, I ended up having to change quite a few things, [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/7f38956b7bc4c51064deb9665acf4a904ae437bf)

It all started with the question of which service layer this logic should belong to. As you said, this is indeed business logic. I tried placing the functions in different spots within the `ProductService`, but I couldn’t find a good place for them. One big reason was that I had linked `Product` and `Option` bidirectionally(an intentional shortcut, but not the most elegant approach) which tightly coupled their creation and made the responsibility unclear. On top of that, I was using `productRepository.save(request.toProduct())`, and the `toProduct` process didn’t even properly create the `Option`. Just... all mixed up.

In the end, I decided to use a `ProductFactory`. It felt wrong for the `Option` domain to create itself from the request, and it also didn’t seem right for the `Product` to be directly responsible for creating options. 

BUT I’m still couldn't move name validation logic to service, , please share if you have any ideas!

I’d like to hear your thoughts.

da-nyee replied:
@miooo0o I replied to it [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15#discussion_r2265325688)!

da-nyee commented on src/main/kotlin/ecommerce/service/CartService.kt:
```diff
@@ -0,0 +1,78 @@
+        memberId: Long,
+        page: Int,
+        size: Int,
+    ): PageImpl<CartItemResponse> {
+        val cart = findCart(memberId)
+        val itemResponses = cart.items.map { CartItemMapper.toResponse(it) }
+        val pageRequest = PageRequest.of(page, size, Sort.by("productName"))
+        val start = pageRequest.offset.toInt()
+        val end = min(start + pageRequest.pageSize, itemResponses.size)
+
+        val pageContent = itemResponses.subList(start, end)
+        return PageImpl<CartItemResponse>(pageContent, pageRequest, itemResponses.size.toLong())
+    }
```
You did well, but it feels a bit complex.

How about this?

```suggestion
    ): Page<CartItemResponse> {
        val cart = findCart(memberId)
        val itemResponses = cart.items.map { CartItemMapper.toResponse(it) }
        val pageRequest = PageRequest.of(page, size, Sort.by("productName"))

        // You need to implement `toResponse()` though.
        return cartRepository.findAll(pageRequest).toResponse()
    }
```

miooo0o replied:
The actual implementation was handled a bit differently.  

Since I wanted to perform pagination directly at the database, I decided to use `CartItemRepository` instead of `CartRepository`.

```kotlin
 ): Page<CartItemResponse>  {
        val pageable = PageRequest.of(page, size, Sort.by("product.name"))
        return cartItemRepository
            .findByCartMemberId(memberId, pageable)
            .map(CartItemMapper::toResponse)
    }
```

applied [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/dd8d9e4da000d5d8ae832f2d115d8f346e187472)!

miooo0o replied:
applied [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/dd8d9e4da000d5d8ae832f2d115d8f346e187472)

da-nyee commented on src/main/kotlin/ecommerce/service/CartService.kt:
```diff
@@ -0,0 +1,78 @@
+
+    fun getPages(
+        memberId: Long,
+        page: Int,
+        size: Int,
```
Have you heard of `no-offset` pagination?
How is it different from the current approach?

miooo0o replied:
Ah, I see.  I completely missed that your question was about `no-offset`!

From what I’ve studied, in my current pagination using `OFFSET` and `LIMIT`, `no-offset` works more like the familiar “Load More” approach.  
If `OFFSET` needs to fetch “10 rows after skipping 1,000,” it will actually read 1,010 rows in total, discard the first 1,000, and then return the remaining 10. The more rows it has to skip, the slower it becomes.

`no-offset`, on the other hand, finds the index of the starting point for the query and only reads as many rows as the page size (like the first page), which makes it much more efficient.

da-nyee replied:
You nailed it 💯

da-nyee commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,77 @@
+        var quantityToRemove = quantity
+        if (existingItem.quantity < quantity) quantityToRemove = existingItem.quantity
+
+        existingItem.quantity -= quantityToRemove
```
What do you think about delegating this responsibility to the `CartItem` rather than accessing it through a getter?

miooo0o replied:
good idea, I guess we implement methods but didn't applied. Now I updated [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/b4287cea5844de51cec630acf782fd5625c4a0cf)

da-nyee commented on src/main/kotlin/ecommerce/service/AuthService.kt:
```diff
@@ -0,0 +1,57 @@
+        val member = memberRepository.findByEmail(payload)
+        if (member == null) {
+            throw NotFoundException("Member not found")
+        }
```
```suggestion
        val member = memberRepository.findByEmail(payload) ?: throw NotFoundException("Member not found")
```

What about using the Elvis operator to make it more idiomatic Kotlin?

miooo0o replied:
Oh, I applied this part in this [commit](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/6f0710336f138c1c2809c51c39da60ec41311e74).

da-nyee commented on src/main/kotlin/ecommerce/service/ProductService.kt:
```diff
@@ -0,0 +1,69 @@
+            ?: throw NotFoundException("Product with name ${productRequest.name} not found")
+    }
+
+    fun read(): List<Product> {
```
```suggestion
    @Transactional(readOnly = true)
    fun read(): List<Product> {
```

You could consider setting `readOnly` when you're only querying data 👀

miooo0o replied:
Thanks for letting me know about the feature. I applied it.

da-nyee commented on src/main/kotlin/ecommerce/service/StatService.kt:
```diff
@@ -0,0 +1,50 @@
+
+        val members =
+            cartItems
+                .map { it.cart.member }
```
Maybe the object could handle it instead of relying on a getter 👀

miooo0o replied:
agree.

```kotlin
            cartItems
                .map { it.getMember() }
                .distinctBy { it.id }
```

Here’s my current approach. I’ll apply the remaining changes, and if I come up with a better idea, I’ll leave another comment.

da-nyee commented on src/main/kotlin/ecommerce/model/Member.kt:
```diff
@@ -0,0 +1,51 @@
+    @Column(name = "role", nullable = false)
+    val role: String = Role.USER.name,
+) {
+    fun validatePassword(password: String) {
```
What do you think about writing unit tests for the models?

miooo0o replied:
Here, `Member.validatePassword` doesn’t actually validate anything. It’s not from my codebase, so if I don’t need to add more validation logic right now, I’d prefer not to touch it until I actually need it. 

I wonder what you think.

da-nyee replied:
Alright, that makes sense!
Let's just leave it as is 👀

da-nyee reviewed (changes requested):
Hi Mina! 🙌

You're doing great so far 👍
There are a few things that could be improved, so I've left some comments — please take a look.

If anything's unclear, just let me know 🙇‍♀️

miooo0o committed:
chore: move DatabaseFixture to test, rename to TestFixture

miooo0o committed:
docs(readme): add feature list based on review feedback

miooo0o committed:
refactor(auth): use the Elvis operator for null checks

miooo0o committed:
refactor(cart, cart-item): delegate quantity adjustment to `CartItem` instead of changing it externally

miooo0o committed:
refactor(product): optimize read() with fetch join and readOnly transaction in repository

- Use fetch join to eagerly load options and prevent N+1 query problem
- Annotate with @Transactional(readOnly = true) for clarity and potential optimization

miooo0o committed:
refactor(product): refactor codebase

- return UpsertStatus and ProductResponse in upsert API
- Return ResponseEntity<ProductResponse> in controller instead of Unit to include more response data
- Reorder parameters for consistency
- non-nullable

miooo0o committed:
refactor(option): using `LAZY` unless eager fetching is required

miooo0o committed:
refactor(option): remove unused X-Page header

miooo0o committed:
refactor(product): prepare service-layer handling for option presence check

miooo0o committed:
refactor(product, option): delegate option binding to ProductFactory

- Removed bidirectional relation setup from Option constructor
  (no longer calls product.options.add(this) in init block)
- Introduced ProductFactory to create Product with pre-built Option entities
- Moved responsibility of binding Option to Product into Product.addOption(s)
  to centralize relation management and maintain domain invariants
- Updated ProductService to use ProductFactory instead of direct entity creation
- Added duplicate option name validation in Product entity

miooo0o committed:
test: applied previous commit changed

- delegate option binding to ProductFactory

miooo0o committed:
docs(readme): move Review.md list to README.md, remove Review.md

miooo0o committed:
refactor(product): move duplication options check logic to Factory

- request.options.distinct().size == request.options.size

miooo0o committed:
refactor(cart): page CartItems by memberId; preload product and cart.member

- Query CartItem via repository instead of loading Cart.items
- Use @EntityGraph(product, cart.member) to avoid N+1
- Keep paging at DB level (no collection fetch join)
- Tests: stub Page via PageImpl/mock; assert content[] structure

miooo0o committed:
chore: remove unused variable

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
Merge remote-tracking branch 'origin/step1' into step1

miooo0o commented:
## Step 1 Refactoring

Hi Dani, thank you for your review.

I tried to apply all your feedback, but as I mentioned in the above comment, I couldn’t really complete this one: "Move product validation to the service layer because the logic belongs to business logic."
### Repository & Service
-  Apply fetch join in `ProductService` and `ProductRepository` to eagerly load options and prevent N+1 queries
-  Replace manual pagination in `CartService` with repository-based pagination
###  Codebase Cleanup
-  Upsert : Return both `UpsertStatus` and `ProductResponse`
-  Change `var` → `val`
-  Remove bidirectional relation setup from `Option` constructor
-  Remove unnecessary commented code (`Product.kt`)
### Factory & Domain Logic
-  Introduce `ProductFactory` to create `Product` with pre-built `Option` entities
-  Move responsibility for binding `Option` to `Product.addOption(s)`
-  Add duplicate option name validation inside `Product` 
### Cart & CartItem
-  Encapsulate quantity adjustment inside `CartItem` instead of modifying it directly in `Cart`
### Test Fixture
-  Rename `DatabaseFixture` → `TestFixture`, Move `DatabaseFixture` to the `test` directory

### Other Refactoring
-  Use Elvis operator for null checks (`AuthService.kt`)
-  Mark read-only methods with `@Transactional(readOnly = true)` (`ProductService.read`)
-  Expose member access through a method in `StatService` instead of nested property access
-  Reconsider adding unit tests for `Member.validatePassword()`

---

### What was the issue during the refactor? 
I can say, I didn’t study fetch type when I wrote that code, and I didn’t consider where/why that logic should be or what it should be with. I checked a few Tecotalk videos and tried to adopt them, but I’m still not sure my understanding is enough or even correct. But it was very interesting.

If there’s a Tecotalk you think would be good for me to watch during this process, could you let me know?
I’ll look for more on my own, but if you have any tips, I’d appreciate it.

Thank you again :)

miooo0o committed:
chore: remove unused methods

da-nyee commented on src/main/kotlin/ecommerce/model/Option.kt:
```diff
@@ -0,0 +1,48 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
+) {
+    @ManyToOne(fetch = FetchType.LAZY)
```
👍

da-nyee commented on src/main/kotlin/ecommerce/service/ProductService.kt:
```diff
@@ -0,0 +1,78 @@
+        return productRepository.save(product).id
+    }
+
+    @Transactional(readOnly = true)
```
👍

da-nyee commented on src/main/kotlin/ecommerce/service/AuthService.kt:
```diff
@@ -0,0 +1,56 @@
+    fun findMember(payload: String): RegisteredMember {
+        val member =
+            memberRepository.findByEmail(payload)
+                ?: throw NotFoundException("Member not found")
```
👍

da-nyee commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,72 @@
+        existingItem: CartItem,
+        quantity: Int,
+    ): CartItem {
+        existingItem.changeQuantityTo(quantity)
+        existingItem.markAsUpdated()
```
👍

da-nyee commented on src/main/kotlin/ecommerce/service/CartService.kt:
```diff
@@ -0,0 +1,75 @@
+        page: Int,
+        size: Int,
+    ): Page<CartItemResponse> {
+        val pageable = PageRequest.of(page, size, Sort.by("product.name"))
+        return cartItemRepository
+            .findByCartMemberId(memberId, pageable)
+            .map(CartItemMapper::toResponse)
```
Looking good 👍

da-nyee commented on src/main/kotlin/ecommerce/service/ProductFactory.kt:
```diff
@@ -0,0 +1,30 @@
+}
+
+@Component
+class DefaultProductFactory : ProductFactory {
```
Well, I think it's a bit over-engineered 😅

What about moving this logic to `ProductService` and injecting `OptionRepository` into it to check for option name duplication? Since you're already using the helper method `addOptions`, mapping both `Product` and `Option` doesn't seem like a big deal. You could also add a helper method in `Option` to bind the `Product`.

da-nyee commented on src/main/kotlin/ecommerce/model/Product.kt:
```diff
@@ -0,0 +1,57 @@
+        return this
+    }
+
+    fun addOption(newOption: Option): Product {
```
This could be `private` 👀

da-nyee reviewed (approved):
Hi Mina! 🙌

Thanks for applying the feedback 🙇‍♀️
I've gone through it and replied to your comments — please check them out when you have time.

Everything's looking good, so let's wrap it up and move on to the next one! 💪
